Jakarta - EE 9 - Infinispan Client IT fails because javax.annotation.Generated is not available

When building the Infinispan Client IT in the `jakarta-rewrite` branch, we get the following failure:

```
Caused by: java.lang.ClassNotFoundException: javax.annotation.Generated
    at java.net.URLClassLoader.findClass (URLClassLoader.java:471)
    at java.lang.ClassLoader.loadClass (ClassLoader.java:589)
    at java.lang.ClassLoader.loadClass (ClassLoader.java:522)
    at org.infinispan.protostream.annotations.impl.processor.AutoProtoSchemaBuilderAnnotationProcessor.addGeneratedClassHeader (AutoProtoSchemaBuilderAnnotationProcessor.java:578)
    at org.infinispan.protostream.annotations.impl.processor.MarshallerSourceCodeGenerator.generateMessageMarshaller (MarshallerSourceCodeGenerator.java:170)
    at org.infinispan.protostream.annotations.impl.processor.MarshallerSourceCodeGenerator.generateMarshaller (MarshallerSourceCodeGenerator.java:58)
    at org.infinispan.protostream.annotations.impl.BaseProtoSchemaGenerator.generateMarshallers (BaseProtoSchemaGenerator.java:203)
    at org.infinispan.protostream.annotations.impl.BaseProtoSchemaGenerator.generateAndRegister (BaseProtoSchemaGenerator.java:174)
    at org.infinispan.protostream.annotations.impl.processor.AutoProtoSchemaBuilderAnnotationProcessor.processClass (AutoProtoSchemaBuilderAnnotationProcessor.java:333)
    at org.infinispan.protostream.annotations.impl.processor.AutoProtoSchemaBuilderAnnotationProcessor.processElement (AutoProtoSchemaBuilderAnnotationProcessor.java:263)
    at org.infinispan.protostream.annotations.impl.processor.AutoProtoSchemaBuilderAnnotationProcessor.process (AutoProtoSchemaBuilderAnnotationProcessor.java:186)
    at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor (JavacProcessingEnvironment.java:980)

```

Several options to fix that:
- test which `@Generated` annotation is available via reflection and use it, instead of blindly use the `javax` one
- switch to a custom `@Generated` annotation hosted in an Infinispan package - this is what Clément did for gRPC
- detect the JDK version and starting from 11 (or use reflection to detect the availability of the annotation), use `javax.annotation.processing.Generated` which is available in the JDK starting 11 instead of the `javax.annotation.Generated` (given only the name of the class is used, it should be an easy switch - obviously you would need to drop the import and use the class names directly)

Note that I have no idea if it's the only problem we will have with Infinispan Client :).

@karesti so I tried to work around it and see if it was the only issue but it's not:

```
Caused by: java.lang.NoClassDefFoundError: javax/transaction/RollbackException
	at org.infinispan.client.hotrod.RemoteCacheManager.<init>(RemoteCacheManager.java:171)
	at org.infinispan.client.hotrod.RemoteCacheManager.<init>(RemoteCacheManager.java:131)
	at io.quarkus.infinispan.client.runtime.InfinispanClientProducer.initialize(InfinispanClientProducer.java:68)
	at io.quarkus.infinispan.client.runtime.InfinispanClientProducer.remoteCacheManager(InfinispanClientProducer.java:314)
	at io.quarkus.infinispan.client.runtime.InfinispanClientProducer_ProducerMethod_remoteCacheManager_b4943dc6282b3982ac4ff16d7d4370041f920ead_Bean.create(Unknown Source)
	at io.quarkus.infinispan.client.runtime.InfinispanClientProducer_ProducerMethod_remoteCacheManager_b4943dc6282b3982ac4ff16d7d4370041f920ead_Bean.get(Unknown Source)
	at io.quarkus.infinispan.client.runtime.InfinispanClientProducer_ProducerMethod_remoteCacheManager_b4943dc6282b3982ac4ff16d7d4370041f920ead_Bean.get(Unknown Source)
	at io.quarkus.it.infinispan.client.CacheSetup_Bean.create(Unknown Source)
	at io.quarkus.it.infinispan.client.CacheSetup_Bean.create(Unknown Source)
	at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:111)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:35)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:32)
	at io.quarkus.arc.impl.LazyValue.get(LazyValue.java:26)
	at io.quarkus.arc.impl.ComputingCache.computeIfAbsent(ComputingCache.java:69)
	at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:32)
	at io.quarkus.it.infinispan.client.CacheSetup_Observer_onStart_a800524f28f3e6295575edd655cd3f1326e59216.notify(Unknown Source)
	at io.quarkus.arc.impl.EventImpl$Notifier.notifyObservers(EventImpl.java:323)
	at io.quarkus.arc.impl.EventImpl$Notifier.notify(EventImpl.java:305)
	at io.quarkus.arc.impl.EventImpl.fire(EventImpl.java:73)
	at io.quarkus.arc.runtime.ArcRecorder.fireLifecycleEvent(ArcRecorder.java:131)
	at io.quarkus.arc.runtime.ArcRecorder.handleLifecycleEvents(ArcRecorder.java:100)
	at io.quarkus.deployment.steps.LifecycleEventsBuildStep$startupEvent1144526294.deploy_0(Unknown Source)
	at io.quarkus.deployment.steps.LifecycleEventsBuildStep$startupEvent1144526294.deploy(Unknown Source)
```

I think we will need a Jakarta version of the Infinispan client. If it's only one artifact to adjust, it should be easy. It could be more painful if there is more.

I'll try to have a closer look and follow up.

--- a/bom/application/pom.xml
+++ b/bom/application/pom.xml
@@ -6086,6 +6086,7 @@
                                 <recipe>io.quarkus.bom.resteasy-spring-web</recipe>
                                 <recipe>io.quarkus.jakarta-json</recipe>
                                 <recipe>io.quarkus.keycloak-admin-client</recipe>
+                                <recipe>io.quarkus.infinispan</recipe>
                             </activeRecipes>
                         </configuration>
                     </plugin>

--- a/extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/InfinispanClientProcessor.java
+++ b/extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/InfinispanClientProcessor.java
@@ -49,13 +49,16 @@ import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
 import io.quarkus.deployment.builditem.ExtensionSslNativeSupportBuildItem;
 import io.quarkus.deployment.builditem.FeatureBuildItem;
 import io.quarkus.deployment.builditem.HotDeploymentWatchedFileBuildItem;
+import io.quarkus.deployment.builditem.NativeImageFeatureBuildItem;
 import io.quarkus.deployment.builditem.SystemPropertyBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.NativeImageConfigBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.NativeImageSecurityProviderBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveClassBuildItem;
+import io.quarkus.deployment.pkg.steps.NativeOrNativeSourcesBuild;
 import io.quarkus.infinispan.client.runtime.InfinispanClientBuildTimeConfig;
 import io.quarkus.infinispan.client.runtime.InfinispanClientProducer;
 import io.quarkus.infinispan.client.runtime.InfinispanRecorder;
+import io.quarkus.infinispan.client.runtime.graal.DisableLoggingFeature;
 import io.quarkus.smallrye.health.deployment.spi.HealthBuildItem;
 
 class InfinispanClientProcessor {
@@ -70,6 +73,11 @@ class InfinispanClientProcessor {
      * The Infinispan client build time configuration.
      */
     InfinispanClientBuildTimeConfig infinispanClient;
+
+    @BuildStep(onlyIf = NativeOrNativeSourcesBuild.class)
+    NativeImageFeatureBuildItem nativeImageFeature() {
+        return new NativeImageFeatureBuildItem(DisableLoggingFeature.class);
+    }
 
     @BuildStep
     InfinispanPropertiesBuildItem setup(ApplicationArchivesBuildItem applicationArchivesBuildItem,

--- a/extensions/infinispan-client/runtime/pom.xml
+++ b/extensions/infinispan-client/runtime/pom.xml
@@ -143,4 +143,28 @@
         </plugins>
     </build>
 
+    <profiles>
+        <profile>
+            <id>jakarta-rewrite</id>
+            <activation>
+                <property>
+                    <name>jakarta-rewrite</name>
+                </property>
+            </activation>
+            <build>
+                <plugins>
+                    <plugin>
+                        <groupId>org.openrewrite.maven</groupId>
+                        <artifactId>rewrite-maven-plugin</artifactId>
+                        <configuration>
+                            <activeRecipes>
+                                <recipe>io.quarkus.infinispan</recipe>
+                            </activeRecipes>
+                        </configuration>
+                    </plugin>
+                </plugins>
+            </build>
+        </profile>
+    </profiles>
+
 </project>

--- a/extensions/infinispan-client/runtime/pom.xml
+++ b/extensions/infinispan-client/runtime/pom.xml
@@ -81,6 +81,10 @@
             <groupId>org.eclipse.microprofile.metrics</groupId>
             <artifactId>microprofile-metrics-api</artifactId>
         </dependency>
+        <dependency>
+            <groupId>io.opentelemetry</groupId>
+            <artifactId>opentelemetry-api</artifactId>
+        </dependency>
 
         <dependency>
             <groupId>org.infinispan</groupId>

--- a/extensions/infinispan-client/runtime/pom.xml
+++ b/extensions/infinispan-client/runtime/pom.xml
@@ -159,6 +159,7 @@
                         <configuration>
                             <activeRecipes>
                                 <recipe>io.quarkus.infinispan</recipe>
+                                <recipe>io.quarkus.fix-infinispan-commons</recipe>
                             </activeRecipes>
                         </configuration>
                     </plugin>

--- a/extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/graal/DisableLoggingFeature.java
+++ b/extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/graal/DisableLoggingFeature.java
@@ -1,0 +1,46 @@
+package io.quarkus.infinispan.client.runtime.graal;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.graalvm.nativeimage.hosted.Feature;
+
+/**
+ * Disables logging during the analysis phase
+ */
+public class DisableLoggingFeature implements Feature {
+
+    private static final String[] CATEGORIES = {
+            "org.infinispan.SECURITY"
+    };
+
+    private final Map<String, Level> categoryMap = new HashMap<>(CATEGORIES.length);
+
+    @Override
+    public void beforeAnalysis(BeforeAnalysisAccess access) {
+        for (String category : CATEGORIES) {
+            Logger logger = Logger.getLogger(category);
+            categoryMap.put(category, logger.getLevel());
+            logger.setLevel(Level.WARNING);
+        }
+    }
+
+    @Override
+    public void afterAnalysis(AfterAnalysisAccess access) {
+        for (String category : CATEGORIES) {
+            Level level = categoryMap.remove(category);
+            if (level != null) {
+                Logger logger = Logger.getLogger(category);
+                logger.setLevel(level);
+            }
+        }
+    }
+
+    @Override
+    public String getDescription() {
+        return "Disables INFO logging during the analysis phase for the " + Arrays.toString(CATEGORIES) + " categories";
+    }
+}

--- a/integration-tests/infinispan-client/pom.xml
+++ b/integration-tests/infinispan-client/pom.xml
@@ -18,11 +18,6 @@
             <artifactId>quarkus-infinispan-client</artifactId>
         </dependency>
 
-        <!-- Allow some basic support for @Inject (required for JAXRS) -->
-        <dependency>
-            <groupId>io.quarkus</groupId>
-            <artifactId>quarkus-arc</artifactId>
-        </dependency>
         <!-- .. and some REST endpoints -->
         <dependency>
             <groupId>io.quarkus</groupId>
@@ -46,49 +41,8 @@
             <artifactId>rest-assured</artifactId>
             <scope>test</scope>
         </dependency>
-        <dependency>
-            <groupId>org.jboss.logging</groupId>
-            <artifactId>commons-logging-jboss-logging</artifactId>
-            <scope>test</scope>
-        </dependency>
-        <dependency>
-            <groupId>org.infinispan</groupId>
-            <artifactId>infinispan-commons-test</artifactId>
-            <scope>test</scope>
-        </dependency>
-        <dependency>
-            <groupId>org.infinispan</groupId>
-            <artifactId>infinispan-core</artifactId>
-            <type>test-jar</type>
-            <scope>test</scope>
-            <exclusions>
-                <!-- jboss-transaction-api is banned and not required for the test -->
-                <exclusion>
-                    <groupId>org.jboss.spec.javax.transaction</groupId>
-                    <artifactId>jboss-transaction-api_1.2_spec</artifactId>
-                </exclusion>
-            </exclusions>
-        </dependency>
-        <dependency>
-            <groupId>org.infinispan</groupId>
-            <artifactId>infinispan-component-annotations</artifactId>
-            <scope>test</scope>
-        </dependency>
 
         <!-- Minimal test dependencies to *-deployment artifacts for consistent build order -->
-        <dependency>
-            <groupId>io.quarkus</groupId>
-            <artifactId>quarkus-arc-deployment</artifactId>
-            <version>${project.version}</version>
-            <type>pom</type>
-            <scope>test</scope>
-            <exclusions>
-                <exclusion>
-                    <groupId>*</groupId>
-                    <artifactId>*</artifactId>
-                </exclusion>
-            </exclusions>
-        </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-infinispan-client-deployment</artifactId>

--- a/jakarta/rewrite.yml
+++ b/jakarta/rewrite.yml
@@ -840,3 +840,15 @@ recipeList:
       newGroupId: org.infinispan
       newArtifactId: infinispan-commons
       newClassifier: jakarta
+---
+type: specs.openrewrite.org/v1beta/recipe
+name: io.quarkus.fix-infinispan-commons
+displayName: Adjust Infinispan version and dependencies
+recipeList:
+  - org.openrewrite.maven.ExcludeDependency:
+      groupId: org.infinispan
+      artifactId: infinispan-commons
+  - org.openrewrite.maven.AddDependencyNoQuestionsAsked:
+      groupId: org.infinispan
+      artifactId: infinispan-commons
+      classifier: jakarta

--- a/jakarta/rewrite.yml
+++ b/jakarta/rewrite.yml
@@ -793,3 +793,50 @@ recipeList:
       oldArtifactId: keycloak-admin-client
       newGroupId: org.keycloak
       newArtifactId: keycloak-admin-client-jakarta
+---
+type: specs.openrewrite.org/v1beta/recipe
+name: io.quarkus.infinispan
+displayName: Adjust Infinispan version and dependencies
+recipeList:
+  - org.openrewrite.maven.ChangePropertyValue:
+      key: infinispan.version
+      newValue: 14.0.0.Dev04
+  - org.openrewrite.maven.ChangePropertyValue:
+      key: infinispan.protostream.version
+      newValue: 14.0.0.CR2
+  - org.openrewrite.maven.ChangeManagedDependencyGroupIdAndArtifactId:
+      oldGroupId: org.infinispan
+      oldArtifactId: infinispan-core
+      newGroupId: org.infinispan
+      newArtifactId: infinispan-core
+      newClassifier: jakarta
+  - org.openrewrite.maven.ChangeManagedDependencyGroupIdAndArtifactId:
+      oldGroupId: org.infinispan
+      oldArtifactId: infinispan-client-hotrod
+      newGroupId: org.infinispan
+      newArtifactId: infinispan-client-hotrod
+      newClassifier: jakarta
+  - org.openrewrite.maven.ChangeManagedDependencyGroupIdAndArtifactId:
+      oldGroupId: org.infinispan
+      oldArtifactId: infinispan-commons
+      newGroupId: org.infinispan
+      newArtifactId: infinispan-commons
+      newClassifier: jakarta
+  - org.openrewrite.maven.ChangeDependencyGroupIdAndArtifactId:
+      oldGroupId: org.infinispan
+      oldArtifactId: infinispan-core
+      newGroupId: org.infinispan
+      newArtifactId: infinispan-core
+      newClassifier: jakarta
+  - org.openrewrite.maven.ChangeDependencyGroupIdAndArtifactId:
+      oldGroupId: org.infinispan
+      oldArtifactId: infinispan-client-hotrod
+      newGroupId: org.infinispan
+      newArtifactId: infinispan-client-hotrod
+      newClassifier: jakarta
+  - org.openrewrite.maven.ChangeDependencyGroupIdAndArtifactId:
+      oldGroupId: org.infinispan
+      oldArtifactId: infinispan-commons
+      newGroupId: org.infinispan
+      newArtifactId: infinispan-commons
+      newClassifier: jakarta

--- a/jakarta/transform.sh
+++ b/jakarta/transform.sh
@@ -218,6 +218,7 @@ update_banned_dependency "build-parent" 'jakarta.xml.bind:jakarta.xml.bind-api' 
 #update_banned_dependency_advanced "build-parent" '<exclude>jakarta.ws.rs:jakarta.ws.rs-api</exclude>' "<exclude>jakarta.ws.rs:jakarta.ws.rs-api</exclude>\n                                            <exclude>org.jboss.spec.javax.ws.rs:jboss-jaxrs-api_2.1_spec</exclude>"
 update_banned_dependency_advanced "build-parent" '<exclude>jakarta.ws.rs:jakarta.ws.rs-api</exclude>' "<exclude>org.jboss.spec.javax.ws.rs:jboss-jaxrs-api_3.0_spec</exclude>\n                                            <exclude>org.jboss.spec.javax.ws.rs:jboss-jaxrs-api_2.1_spec</exclude>"
 update_banned_dependency_advanced "build-parent" '<exclude>jakarta.json:jakarta.json-api</exclude>' "<exclude>jakarta.json:jakarta.json-api</exclude>\n                                            <exclude>org.glassfish:jakarta.json</exclude>"
+update_banned_dependency_advanced "build-parent" '<exclude>org.glassfish:javax.el</exclude>' "<exclude>org.glassfish:javax.el</exclude>\n                                            <exclude>org.glassfish:jakarta.el</exclude>"
 sed -i 's@<!-- Exclude jakarta.activation-api as the implementation contains it -->@<!-- Exclude com.sun.activation:jakarta.activation as we switched to Angus Activation -->@g' 'build-parent/pom.xml'
 update_banned_dependency "build-parent" 'jakarta.activation:jakarta.activation-api' 'com.sun.activation:jakarta.activation'
 

--- a/jakarta/transform.sh
+++ b/jakarta/transform.sh
@@ -334,7 +334,6 @@ fi
 # - Infinispan uses the @javax.annotation.Generated annotation in code generation and it's not available
 # - Confluent registry client doesn't have a version supporting Jakarta packages
 sed -i 's@<module>kafka-avro</module>@<!-- <module>kafka-avro</module> -->@g' integration-tests/pom.xml
-sed -i 's@<module>infinispan-client</module>@<!-- <module>infinispan-client</module> -->@g' integration-tests/pom.xml
 
 ./mvnw -B clean install -f integration-tests -DskipTests -DskipITs
 ./mvnw -B clean install -f tcks -DskipTests -DskipITs

--- a/test-framework/infinispan-client/pom.xml
+++ b/test-framework/infinispan-client/pom.xml
@@ -72,6 +72,10 @@
                     <groupId>org.wildfly.security</groupId>
                     <artifactId>*</artifactId>
                 </exclusion>
+                <exclusion>
+                    <groupId>org.latencyutils</groupId>
+                    <artifactId>LatencyUtils</artifactId>
+                </exclusion>
             </exclusions>
         </dependency>
         <dependency>

